// aws-cdk/lib/api/bootstrap-environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const deploy_stack_1 = require("./deploy-stack");
// tslint:disable:max-line-length
exports.BUCKET_NAME_OUTPUT = 'BucketName';
exports.BUCKET_DOMAIN_NAME_OUTPUT = 'BucketDomainName';
async function bootstrapEnvironment(environment, aws, toolkitStackName, roleArn) {
    const synthesizedStack = {
        environment,
        metadata: {},
        template: {
            Description: "The CDK Toolkit Stack. It cas created by `cdk bootstrap` and manages resources necessary for managing your Cloud Applications with AWS CDK.",
            Resources: {
                StagingBucket: {
                    Type: "AWS::S3::Bucket",
                    Properties: {
                        AccessControl: "Private",
                        BucketEncryption: { ServerSideEncryptionConfiguration: [{ ServerSideEncryptionByDefault: { SSEAlgorithm: "aws:kms" } }] }
                    }
                }
            },
            Outputs: {
                [exports.BUCKET_NAME_OUTPUT]: {
                    Description: "The name of the S3 bucket owned by the CDK toolkit stack",
                    Value: { Ref: "StagingBucket" },
                    Export: { Name: "CDKToolkit:BucketName" }
                },
                [exports.BUCKET_DOMAIN_NAME_OUTPUT]: {
                    Description: "The domain name of the S3 bucket owned by the CDK toolkit stack",
                    Value: { "Fn::GetAtt": ["StagingBucket", "DomainName"] },
                    Export: { Name: "CDKToolkit:BucketDomainName" }
                }
            }
        },
        name: toolkitStackName,
    };
    return await deploy_stack_1.deployStack({ stack: synthesizedStack, sdk: aws, roleArn });
}
exports.bootstrapEnvironment = bootstrapEnvironment;
